contracts/DivviRegistry.sol

```diff
@@ -0,0 +1,244 @@
+
+  // Events
+  event RewardsEntityRegistered(address indexed entity, bool requiresApproval);
+  event RequiresApprovalForRewardsAgreements(
+    address indexed entity,
+    bool requiresApproval
+  );
+  event RewardsAgreementRegistered(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer
+  );
+  event RewardsAgreementApproved(
+    address indexed rewardsProvider,
+    address indexed rewardsConsumer
+  );
+  event ReferralRegistered(
+    address indexed user,
+    address indexed rewardsConsumer,
+    address indexed rewardsProvider
+  );
+
+  // Errors
+  error InvalidEntityAddress(address entity);
+  error EntityAlreadyExists(address entity);
+  error EntityDoesNotExist(address entity);
+  error AgreementAlreadyExists(address provider, address consumer);
+  error AgreementDoesNotExist(address provider, address consumer);
+  error ProviderRequiresApproval(address provider);
+  error UserAlreadyReferred(address user, address provider);
+  error MissingReferralRegistrarRole(address account);
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(address owner, uint48 transferDelay) public initializer {
+    __AccessControlDefaultAdminRules_init(transferDelay, owner);
+    __UUPSUpgradeable_init();
+  }
+
+  function _authorizeUpgrade(
+    address
+  ) internal override onlyRole(DEFAULT_ADMIN_ROLE) {} // solhint-disable-line no-empty-blocks
+
+  /**
+   * @notice Modifier to ensure an entity exists
+   * @param entity The entity address to check
+   */
+  modifier entityExists(address entity) {
+    if (!_entities[entity]) {
+      revert EntityDoesNotExist(entity);
+    }
+    _;
+  }
+
+  /**
+   * @notice Register a new rewards entity
+   * @param entity The entity address to register
+   * @param requiresApproval Whether the entity requires approval for agreements
+   */
+  function registerRewardsEntity(
+    address entity,
+    bool requiresApproval
+  ) external {
+    if (entity == address(0)) {
+      revert InvalidEntityAddress(entity);
+    }
+
+    if (_entities[entity]) {
+      revert EntityAlreadyExists(entity);
+    }
+
+    _entities[entity] = true;
+    _requiresApproval[entity] = requiresApproval;
+    emit RewardsEntityRegistered(entity, requiresApproval);
+  }
+
+  /**
+   * @notice Set whether a Rewards Entity requires approval for agreements
+   * @param requiresApproval Whether the entity requires approval
+   */
+  function setRequiresApprovalForRewardsAgreements(
+    bool requiresApproval
+  ) external entityExists(msg.sender) {
+    _requiresApproval[msg.sender] = requiresApproval;
+    emit RequiresApprovalForRewardsAgreements(msg.sender, requiresApproval);
+  }
+
+  /**
+   * @notice Register a Rewards Consumer - Rewards Provider relationship
+   * @dev Should be called by the Rewards Consumer
+   * @param rewardsProvider The provider entity address
+   */
+  function registerRewardsAgreement(
```

Comment:
If a provider requires approvals, how is the consumer supposed to ever register the agreement? I don't see a way in the contract for e.g., a consumer to "propose" an agreement with a provider that requires approvals, such that the provider can later approve it.